### PR TITLE
Custom backgroundSize for image layouts (#798)

### PR DESCRIPTION
**Pull request by alpcoker**
fixes https://github.com/slidevjs/slidev/issues/798

With this additional prop to image layouts (image, image-left, image-right), we can now able give custom css [background-size](https://developer.mozilla.org/en-US/docs/Web/CSS/background-size). The default will remain `cover` 

My main motivation was to be able to use `contain`. Because I wasn't able to fit vertical photographs, let's cover all the different needs.  

Example usages:

```
---
layout: image
image: https://iili.io/HzdAhJV.png
backgroundSize: contain
---
```

```
---
layout: image-left
image: https://iili.io/HzdAhJV.png
backgroundSize: 20em 70%
---
```

### CODE REVIEW

**Comment by antfu** [comment]
I honestly feel this is a bit too specially handled. When you want to customize it, there are many more other styles to configure as well. I'd suggest we expose classes or styles to the image. Maybe:

```md
---
imageStyle:
  backgroundPosition: center
  backgroundSize: etc
---
```

**Comment by alpcoker** [comment]
I love the simplicity of the layout `image[-left|-right].vue` by just setting the image URL. The only problem is that the existing `background-size: cover` is not suitable for some usages because it scales.

I also agree that this extra property seems very specific, but I think the `background-size` property would be the most used `background-*` property after `background-image`. I am not sure that other properties would be used that much.

- I tried your opinion for an ability to set other background properties with a named `image-custom` layout.
- I also wanted to try another layout that's `background-size: contain` as a static. The drawback is to create for also left & right layout too.

To compare & discuss 3 options temporarily, I have changed the demo project's `slide.md`

**Comment by alpcoker** [comment]
Hello @antfu. As I see conflicts in my PR, I rebased and for the demo I added `pnpm demo:feat` (just image layout demos created in slides-feat.md) to be easily review. I also added the slide screenshots.

File: demo/starter/slides-feat.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+---
```
Comment (antfu) [inline demo/starter/slides-feat.md:1]: The starter is meant to be an example showcase and a playground. For tests like this it's better to move to test fixtures.

File: packages/client/layouts/image-custom.vue [hunk 2]
```diff
@@ -0,0 +1,50 @@
+<script setup lang="ts">
```
Comment (antfu) [inline packages/client/layouts/image-custom.vue:1]: I am not sure about this layout. I am concerned about the slope effect that we will end up with all CSS rules. I'd recommend creating your own local layouts for your specific needs.

**Review by antfu — CHANGES_REQUESTED** [review]

File: packages/client/layouts/image-custom.vue (hunk 2 above)
Comment (alpcoker) [reply packages/client/layouts/image-custom.vue]: Sorry for accidentally deleting the branch, fix that & I have [opened new PR](https://github.com/slidevjs/slidev/pull/862) so we can continue from there.

File: demo/starter/slides-feat.md (hunk 1 above)
Comment (alpcoker) [reply demo/starter/slides-feat.md]: Sorry for accidentally deleting the branch, fix that & I have [opened new PR](https://github.com/slidevjs/slidev/pull/862) so we can continue from there.

**Pull request closed without merging**